Range-separated density fitting for PBC calculations

hongzhouye opened the pull request:
This PR adds the implementation of the range-separated gaussian density fitting (RSGDF) developed by @tberkel and me as detailed in the following two manuscripts:
1. Ye, and Berkelbach _J. Chem. Phys._ **154**, 131104 (2021) DOI: 10.1063/5.0046617
2. Ye, and Berkelbach _J. Chem. Phys._ **155**, 124106 (2021) DOI: 10.1063/5.0064151

It is implemented as a derived class of `pbc.df.df.GDF` and can be used in almost the same way as `GDF`. E.g.,
```Python
mf = scf.KRHF(cell, kpts).rs_density_fit()
```

We did a decent amount of tests that cover (huge thanks to @btglau, @verena-neufeld, @zhutianyu1991, @Majokux, and @cnyeh for running tests and providing useful feedbacks on preliminary versions of this module)
- different materials and basis sets
- different levels of theory including SCF, MP2, CCSD, EOM-CCSD, 
- k-point sampling vs gamma point with big supercell

against the current `GDF` in PySCF. The agreement between RSGDF and GDF is in general perfect, with a typical difference in energy 10^-8 Eh per cell when using the default `cell.precision`. The speedup of RSGDF warm-up over GDF is typically 10 ~ 100x when measured in terms of CPU time or 5 ~ 10x in wall time depending on the system/basis sets and number of CPU cores being used. The smaller speedup in wall time is due to the I/O bound affecting RSGDF more (because it's faster).

verena-neufeld commented on pyscf/pbc/df/rsdf.py:
```diff
@@ -0,0 +1,703 @@
+    # overwriting.  It often leads to the cderi_file ~2 times larger than the
+    # necessary size.  For now, dumping the DF integral intermediates to a
+    # separated temporary file can avoid this issue.  The DF intermediates may
+    # be terribly huge. The temporary file should be placed in the same disk
```
Why should it be placed on the same disk?

hongzhouye replied:
these lines were copied from GDF

verena-neufeld commented on pyscf/pbc/df/rsdf.py:
```diff
@@ -0,0 +1,703 @@
+        if cell.dimension < 3:
+            raise NotImplementedError("RSGDF for low-dimensional systems are not available yet. We recommend using cell.dimension=3 with large vacuum.")
+
+        # if True and kpts are gamma-inclusive, RSDF will use the bvk cell trick for computing both j3c_SR and j3c_LR. If kpts are not gamma-inclusive, this attribute will be ignored.
```
This is quite a long line

hongzhouye replied:
fixed.

verena-neufeld commented on pyscf/pbc/df/rsdf.py:
```diff
@@ -0,0 +1,703 @@
+        self.precision_R = cell.precision * 1e-2
+        self.precision_G = cell.precision
+
+        # One of {omega, npw_max} must be provided, and the other will be deduced automatically from it. The priority when both are given is omega > npw_max.
```
This is quite a long line

hongzhouye replied:
fixed.

verena-neufeld commented on pyscf/pbc/df/rsdf.py:
```diff
@@ -0,0 +1,703 @@
+
+        # omega and mesh for j2c. Since j2c is to be inverted, it is desirable to use (1) a "fixed" omega for reproducibility, and (2) a higher precision to minimize any round-off error caused by inversion.
+        # The extra computational cost due to the higher precision is negligible compared to the j3c build.
+        # FOR EXPERTS: if you want omega_j2c to be the same as the omega used for j3c build, set omega_j2c to be any negative number.
```
is this safe? would it make sense to allow a new variable the user can pass?

verena-neufeld commented on pyscf/pbc/df/rsdf_helper.py:
```diff
@@ -0,0 +1,1229 @@
+def get_2c2e_Rcut(bas_lst, cellvol, omega, precision, Rprec=1.,
+                  lmp=True, lasympt=True,
+                  eta_correct=True, R_correct=False, vol_correct=False):
+    r""" Given a list of pgto by "bas_lst", determine the cutoff radii for j2c lat sum s.t. the truncation error drops below "precision". j2c is estimated as
```
This line is quite long

verena-neufeld commented on pyscf/pbc/df/rsdf_helper.py:
```diff
@@ -0,0 +1,1229 @@
+            "4c"-mode:  Q = 2-norm[(ab|ab)]^(1/2)
+    """
+    def get1ctr(bas_lst):
+        """ For a shell consists of multiple contracted GTOs, keep only the one with the greatest weight on the most diffuse primitive GTOs (since others are likely core orbitals).
```
This is quite a long line

verena-neufeld commented on pyscf/pbc/df/rsdf_helper.py:
```diff
@@ -0,0 +1,1229 @@
+    return Qs
+def get_schwartz_dcut(bas_lst, cellvol, omega, precision, r0=None, safe=True,
+                      vol_correct=False):
+    """ Given a list of basis, determine cutoff radius for the Schwartz Q between each unique shell pair to drop below "precision". The Schwartz Q is define:
```
This is quite a long line

hongzhouye replied:
fixed.

verena-neufeld commented on pyscf/pbc/df/rsdf_helper.py:
```diff
@@ -0,0 +1,1229 @@
+        def fcheck(R):
+            mol._env[mol._atm[1,mol_gto.PTR_COORD]] = R
+            I = get_norm(fintor_sreri(mol, intor, shls_slice, omega, safe))**0.5
+# >>>>>>> debug block
```
What do you mean with debug block?

hongzhouye replied:
for experts :D

verena-neufeld commented on pyscf/pbc/df/rsdf_helper.py:
```diff
@@ -0,0 +1,1229 @@
+                    dj = ei/eij * (dij + extij)
+                where "extij" is the extent of orbital pair ij.
+
+        Similar to :func:`get_2c2e_Rcut`, the estimator is multiplied by factor of eta and/or 1/R if "eta_correct" and/or "R_correct" are set to True.
```
This is quite a long line

hongzhouye replied:
fixed.

verena-neufeld commented on pyscf/pbc/df/rsdf_jk.py:
```diff
@@ -0,0 +1,52 @@
+import copy
+
+def density_fit(mf, auxbasis=None, mesh=None, with_df=None):
+    '''Generte density-fitting SCF object
```
Generate

hongzhouye replied:
Copied from `df_jk.py` so I will leave this feature typo for now.

hongzhouye replied:
fixed

verena-neufeld commented on pyscf/pbc/df/test/test_rsdf.py:
```diff
@@ -0,0 +1,80 @@
+    def test_h2_gamma(self):
+        mf = pscf.KRHF(cell).rs_density_fit()
+        mf.kernel()
+        self.assertAlmostEqual(mf.e_tot, -1.0430635251, 6)
```
Is `6` the accuracy you'd expect with the default RSGDF settings?

hongzhouye replied:
Changed to 7. RSDF computes integrals to be `cell.precision` which by default is 1e-8. It's not clear exactly how that is translated to the precision in SCF energy but a bit larger than that is expected in general.

verena-neufeld reviewed (commented):
Hi Hong-Zhou! Super cool stuff, I'll continue testing this :D. For now, I have started with a very superficial review of some super minor comments. Feel free to ignore as many of these as you like. I think some tests are currently failing here due to flake8.

sunqm commented:
In nr3c_k and other `fill` functions, the short range Coulomb integrals are computed as `full_integrals - long_range_integrals` . They can be computed directly using the releases of libcint 4.4.4 and newer. Could you merge the master and test whether it gives different output and changes performance?

hongzhouye commented on pyscf/pbc/df/rsdf.py:
```diff
@@ -0,0 +1,703 @@
+where the SR and LR integrals correpond to using the following potentials
+    g_SR(r_12;omega) = erfc(omega * r_12) / r_12
+    g_LR(r_12;omega) = erf(omega * r_12) / r_12
+The SR integrals are evaluated in real space using a lattice summation, while the LR integrals are evaluated in reciprocal space with a plane wave basis.
```
fixed.

hongzhouye commented on pyscf/pbc/df/rsdf.py:
```diff
@@ -0,0 +1,703 @@
+            j2c = scipy.linalg.cholesky(j2c, lower=True)
+            j2ctag = 'CD'
+        except scipy.linalg.LinAlgError:
+            #msg =('===================================\n'
```
these lines are copied from GDF so I kept them unchanged.

hongzhouye commented:
@sunqm Hi Qiming, as we discussed last time, I
1. fixed the get_lattice_Ls bug in [this commit](https://github.com/pyscf/pyscf/pull/1034/commits/84106178bdeb6a8807a21e91478a8d45bd443497).
2. added tests in [this commit](https://github.com/pyscf/pyscf/pull/1034/commits/213cd38faddc938791ade775bcf686219da96063). They are mostly taken from the corresponding tests for GDF, with known values being modified slightly to account for the small difference between RSDF and GDF (1e-6 ~ 1e-5 for most matrix elements).
3. added examples in [this commit](https://github.com/pyscf/pyscf/pull/1034/commits/f20779b9a78e3f9bdf74a22b4e226115a19e62b2).

I hope these changes make the PR ready for a final review.

sunqm commented on pyscf/pbc/df/rsdf_helper.py:
```diff
@@ -0,0 +1,1297 @@
+#!/usr/bin/env python
```
Many functions in this file are needed only by this module. Can you prefix them with `_` to indicate they are private functions? For functions which may be used by other modules, please add docs for all arguments.

hongzhouye replied:
fixed in [this commit](https://github.com/pyscf/pyscf/pull/1034/commits/94585b16b1987beda31e273b0e5e2bc003e76f01)

sunqm commented on pyscf/pbc/df/rsdf_helper.py:
```diff
@@ -0,0 +1,1297 @@
+    return cbins
+def get_3c2e_Rcuts_for_d(mol, auxmol, ish, jsh, dij, cellvol, omega, precision,
+                         fac_type, Qij, Rprec=1,
+                         vol_correct=False, eta_correct=True, R_correct=True):
```
What are fac_type, Qij, Rprec, vol_correct, eta_correct, R_correct? Please add docs for these arguments.

hongzhouye replied:
fixed in [this commit](https://github.com/pyscf/pyscf/pull/1034/commits/94585b16b1987beda31e273b0e5e2bc003e76f01)

sunqm commented on pyscf/pbc/df/rsdf_helper.py:
```diff
@@ -0,0 +1,1297 @@
+        expkL.ctypes.data_as(ctypes.c_void_p),
+        (ctypes.c_int*4)(i0, i1, j0, j1),
+        ao_loc.ctypes.data_as(ctypes.c_void_p), cintopt,
+        refuniqshl_map.ctypes.data_as(ctypes.c_void_p),
```
ensure dtype of refuniqshl_map. (call np.asarray(..., dtype=) before calling the C function)

hongzhouye replied:
the dtype of `refuniqshl_map` is already set to numpy.int32 in the function generating it `_get_refuniq_map`. See L171.

Do I need to add another `np.asarray(..., dtype=)` right before calling the C function in `intor_j2c`?

sunqm commented:
`rs_density_fit` is supposed to deliver a 3-index tensor the same to the one from `density_fit`, right? When using the default parameters, which one is not tight enough that causes the small differences between the two?

hongzhouye commented on pyscf/lib/pbc/fill_ints_sr.c:
```diff
@@ -0,0 +1,2002 @@
+#include "cint.h"
+#include "vhf/fblas.h"
+#include "np_helper/np_helper.h"
+
```
fixed in [this commit](https://github.com/pyscf/pyscf/pull/1034/commits/669cba9c9c1315d7f2269c161bafbd01bf038cc7)

hongzhouye commented on pyscf/pbc/df/rsdf_helper.py:
```diff
@@ -0,0 +1,1297 @@
+    """ Given a list of basis ("bas_lst") and auxiliary basis ("auxbas_lst"),
+    determine the cutoff radius for
+        2-norm( (k|v_SR(omega)|ij) ) < precision
+    where i and j shls are separated by d specified by "dijs_lst".
```
fixed in [this commit](https://github.com/pyscf/pyscf/pull/1034/commits/94585b16b1987beda31e273b0e5e2bc003e76f01)

hongzhouye commented on pyscf/pbc/df/rsdf_helper.py:
```diff
@@ -0,0 +1,1297 @@
+                (ctypes.c_int*6)(*shls_slice),
+                ao_loc.ctypes.data_as(ctypes.c_void_p),
+                cintopt,
+                shlpr_mask.ctypes.data_as(ctypes.c_void_p),  # shlpr_mask
```
`shlpr_mask` is set to `numpy.int8`; see L1053
`refuniqshl_map` and `auxuniqshl_map` are set to `numpy.int32` in the function generating them;  see L171
`dijs_loc` is set to `numpy.int32` when generated; see L1007
Other arrays are all double (or like `atm`, `bas` etc directly taken from `cell`).

hongzhouye commented:
@sunqm I *think* the main part of the small difference comes from using a different auxiliary basis. RSDF uses the default while GDF re-normalizes the fitting functions.
